### PR TITLE
Hooks for Akka Persistence Typed lifecycle

### PR DESCRIPTION
**Pull request by chbatey**

### CODE REVIEW

**Comment by chbatey** [comment]
@raboof FYI ^

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala [hunk 1]
```diff
@@ -69,6 +68,17 @@ private[akka] final class ReplayingEvents[C, E, S](
   import ReplayingEvents.ReplayingState
 
   replayEvents(state.seqNr + 1L, state.toSeqNr)
+  onRecoveryStart(setup.context)
+
+  @InternalStableApi
+  def onRecoveryStart(@unused context: ActorContext[_]): Unit = {}
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala]: our normal convention is to use `()` for such empty impl

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala [hunk 2]
```diff
@@ -167,11 +177,18 @@ private[akka] final class ReplayingEvents[C, E, S](
...
    */
   private def onRecoveryFailure(cause: Throwable, event: Option[Any]): Behavior[InternalProtocol] = {
+    onRecoveryFailed(setup.context, cause, event)
     setup.onSignal(state.state, RecoveryFailed(cause), catchAndLog = true)
     setup.cancelRecoveryTimer()
     tryReturnRecoveryPermit("on replay failure: " + cause.getMessage)
-
+    if (setup.log.isDebugEnabled) {
+      setup.log.debug(
+        "Recovery failure for persitsenceId {} after {} nanos",
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala]: worth formatting those nanos with pretty duration or FiniteDuration.toCoarsest ?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala [hunk 3]
```diff
@@ -186,7 +203,14 @@ private[akka] final class ReplayingEvents[C, E, S](
 
   private def onRecoveryCompleted(state: ReplayingState[S]): Behavior[InternalProtocol] =
     try {
+      onRecoveryComplete(setup.context)
       tryReturnRecoveryPermit("replay completed successfully")
+      if (setup.log.isDebugEnabled) {
+        setup.log.debug(
+          "Recovery for persistenceId {} took {} nanos",
```
Comment (patriknw) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala]: `[{}]` for consistency with most such logging

**Review by patriknw — APPROVED** [review]
LGTM, minor nits

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala [hunk 4]
```diff
@@ -167,11 +177,18 @@ private[akka] final class ReplayingEvents[C, E, S](
...
+    onRecoveryFailed(setup.context, cause, event)
     setup.onSignal(state.state, RecoveryFailed(cause), catchAndLog = true)
     setup.cancelRecoveryTimer()
     tryReturnRecoveryPermit("on replay failure: " + cause.getMessage)
-
+    if (setup.log.isDebugEnabled) {
+      setup.log.debug(
+        "Recovery failure for persitsenceId {} after {} nanos",
+        setup.persistenceId,
+        System.nanoTime() - state.recoveryStartTime)
```
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala]: Nanos seems like a quite bad unit here? Pretty print time with `akka.util.PrettyDuration`?

**Review by johanandren — APPROVED** [review]

**Pull request merged**